File: sdk-python/obsvr/strict_receipt_coordinator_v2.py
"""Two-phase strict v2 receipt coordinator with tenant-bound state."""

from __future__ import annotations

import copy
import os
import threading
import uuid
from typing import Any, Callable, Dict, List, Optional

from .strict_receipt_coordinator_v2_support import (
    build_coordinator_context_v2,
    build_intent_policy_v2,
    context_input_from_v2_document,
    decision_v2_fingerprint,
    evaluate_intent_alignment_v2,
    resolution_v2_fingerprint,
    sign_decision_v2,
    sign_resolution_v2,
    timeout_v2_fingerprint,
    v2_canonical_hash,
    v2_hash,
    v2_integer,
    v2_text,
    validate_resolution_v2,
)
from .strict_receipt_prepared_state import PreparedReceiptState
from .strict_receipt_v2 import strict_receipt_v2_key_id


class StrictReceiptCoordinatorV2Error(ValueError):
    """Strict v2 coordination cannot proceed safely."""


class StrictReceiptCoordinatorV2:
    """Prepare signed v2 receipts, then advance only after durable admission."""

    def __init__(
        self,
        *,
        signer: Any,
        policy: Dict[str, Any],
        sdk_language: str,
        sdk_version: str,
        tenant_id: str,
        session_id: str,
        clock: Callable[[], int],
        defer_ttl_ms: int,
        approval_verifier: Callable[[Any, Dict[str, Any]], Dict[str, Any]],
        include_public_key: bool = True,
        pid: Callable[[], int] = os.getpid,
        prepared_token_factory: Callable[[], str] = lambda: str(uuid.uuid4()),
    ) -> None:
        if sdk_language != "python":
            raise StrictReceiptCoordinatorV2Error("sdk_language must be python")
        self._signer = signer
        self._policy = build_intent_policy_v2(policy)
        self._sdk_version = v2_text(sdk_version, "sdk_version")
        self._tenant_id = v2_text(tenant_id, "tenant_id")
        self._session_id = v2_text(session_id, "session_id")
        self._clock = clock
        self._defer_ttl_ms = v2_integer(defer_ttl_ms, "defer_ttl_ms")
        if self._defer_ttl_ms == 0:
            raise StrictReceiptCoordinatorV2Error("defer_ttl_ms must be positive")
        if not callable(approval_verifier):
            raise StrictReceiptCoordinatorV2Error(
                "approval_verifier must be a function"
            )
        self._approval_verifier = approval_verifier
        self._include_public_key = include_public_key
        self._pid_source = pid
        self._owner_pid = v2_integer(pid(), "pid")
        strict_receipt_v2_key_id(signer.raw_public_key)
        self._prepared_state = PreparedReceiptState(prepared_token_factory)
        self._sequence = 0
        self._last_receipt_hash: Optional[str] = None
        self._last_timestamp: Optional[int] = None
        self._prior_actions: List[Dict[str, Any]] = []
        self._suspended: Dict[str, Dict[str, Any]] = {}
        self._resolved: set[str] = set()
        self._action_ids: set[str] = set()
        self._approval_requests: Dict[str, str] = {}
        self._lock = threading.RLock()

    def prepare_decision(
        self,
        *,
        context: Dict[str, Any],
        base_result: Dict[str, Any],
        policy_version: str,
        rule_ids: List[str],
        action_id: str,
    ) -> Dict[str, Any]:
        with self._lock:
            self._ensure_process()
            input_value = {
                "context": context,
                "base_result": base_result,
                "policy_version": policy_version,
                "rule_ids": rule_ids,
                "action_id": action_id,
            }
            action_id = v2_text(action_id, "action_id")
            fingerprint = decision_v2_fingerprint(
                input_value=input_value,
                tenant_id=self._tenant_id,
                session_id=self._session_id,
            )
            if action_id in self._action_ids:
                raise StrictReceiptCoordinatorV2Error("action_id is already committed")
            retry = self._prepared_state.retry(fingerprint, "decision")
            if retry is not None:
                return copy.deepcopy(retry)
            normalized = build_coordinator_context_v2(
                context, self._session_id, self._prior_actions
            )
            base = copy.deepcopy(base_result)
            evaluation = evaluate_intent_alignment_v2(
                context=normalized, base_result=base, policy=self._policy
            )
            timestamp, clamped = self._allocate_timestamp()
            receipt = sign_decision_v2(
                action_id=action_id,
                tenant_id=self._tenant_id,
                session_id=self._session_id,
                sequence=self._sequence + 1,
                timestamp=timestamp,
                clamped=clamped,
                previous_hash=self._last_receipt_hash,
                sdk_version=self._sdk_version,
                signer=self._signer,
                include_public_key=self._include_public_key,
                context=normalized,
                base_result=base,
                evaluation=evaluation,
                policy_version=policy_version,
                rule_ids=rule_ids,
                defer_ttl_ms=self._defer_ttl_ms,
            )
            request_id = (
                receipt["body"].get("suspension", {}).get("approval_request_id")
            )
            if request_id in self._approval_requests:
                raise StrictReceiptCoordinatorV2Error(
                    "approval_request_id is already pending"
                )
            result = {"evaluation": evaluation, "receipt": receipt}
            prepared = self._prepared_state.prepare(
                fingerprint=fingerprint,
                receipt_hash=receipt["receipt_hash"],
                kind="decision",
                value=result,
                commit=lambda: self._commit_decision(result, normalized, base),
            )
            return copy.deepcopy(prepared)

    def prepare_resolution(
        self,
        *,
        suspended_receipt_hash: str,
        method: str,
        context: Dict[str, Any],
        base_result: Dict[str, Any],
        policy_version: str,
        rule_ids: List[str],
        resolver_principal_id: Optional[str] = None,
        resolution_source_hash: Optional[str] = None,
        approval_evidence_value: Any = None,
    ) -> Dict[str, Any]:
        with self._lock:
            self._ensure_process()
            if method == "expired":
                raise StrictReceiptCoordinatorV2Error(
                    "expired suspensions must use prepare_timeout"
                )
            input_value = {
                "suspended_receipt_hash": suspended_receipt_hash,
                "method": method,
                "context": context,
                "base_result": base_result,
                "policy_version": policy_version,
                "rule_ids": rule_ids,
                "resolver_principal_id": resolver_principal_id,
                "resolution_source_hash": resolution_source_hash,
                "approval_evidence": approval_evidence_value,
            }
            fingerprint = resolution_v2_fingerprint(
                input_value=input_value,
                tenant_id=self._tenant_id,
                session_id=self._session_id,
            )
            retry = self._prepared_state.retry(fingerprint, "resolution")
            if retry is not None:
                return copy.deepcopy(retry)
            receipt_hash = v2_hash(suspended_receipt_hash, "suspended_receipt_hash")
            pending = self._pending(receipt_hash)
            normalized = build_coordinator_context_v2(
                context, self._session_id, self._prior_actions
            )
            base = copy.deepcopy(base_result)
            evaluation = evaluate_intent_alignment_v2(
                context=normalized, base_result=base, policy=self._policy
            )
            timestamp, clamped = self._allocate_timestamp()
            evidence = validate_resolution_v2(
                input_value=input_value,
                pending=pending,
                context=normalized,
                base=base,
                evaluation=evaluation,
                timestamp=timestamp,
                approval_verifier=self._approval_verifier,
                approval_requests=self._approval_requests,
            )
            index = next(
                (
                    index
                    for index, item in enumerate(self._prior_actions)
                    if item["receipt_hash"] == receipt_hash
                ),
                None,
            )
            if index is None:
                raise StrictReceiptCoordinatorV2Error(
                    "suspended action summary is missing"
                )
            receipt = sign_resolution_v2(
                prior=pending["receipt"],
                method=method,
                principal_id=evidence["principal_id"],
                source_hash=evidence["source_hash"],
                tenant_id=self._tenant_id,
                session_id=self._session_id,
                sequence=self._sequence + 1,
                timestamp=timestamp,
                clamped=clamped,
                previous_hash=self._last_receipt_hash,
                sdk_version=self._sdk_version,
                signer=self._signer,
                include_public_key=self._include_public_key,
                context=normalized,
                base_result=base,
                evaluation=evaluation,
                policy_version=policy_version,
                rule_ids=rule_ids,
            )
            prepared = self._prepared_state.prepare(
                fingerprint=fingerprint,
                receipt_hash=receipt["receipt_hash"],
                kind="resolution",
                value=receipt,
                commit=lambda: self._commit_resolution(receipt, receipt_hash, index),
            )
            return copy.deepcopy(prepared)

    def prepare_timeout(
        self,
        *,
        suspended_receipt_hash: str,
        policy_version: str,
        rule_ids: List[str],
    ) -> Dict[str, Any]:
        with self._lock:
            self._ensure_process()
            input_value = {
                "suspended_receipt_hash": suspended_receipt_hash,
                "policy_version": policy_version,
                "rule_ids": rule_ids,
            }
            fingerprint = timeout_v2_fingerprint(
                input_value=input_value,
                tenant_id=self._tenant_id,
                session_id=self._session_id,
            )
            retry = self._prepared_state.retry(fingerprint, "timeout")
            if retry is not None:
                return copy.deepcopy(retry)
            receipt_hash = v2_hash(suspended_receipt_hash, "suspended_receipt_hash")
            pending = self._pending(receipt_hash)
            suspension = pending["receipt"]["body"]["suspension"]
            normalized = build_coordinator_context_v2(
                context_input_from_v2_document(pending["context"]),
                self._session_id,
                self._prior_actions,
            )
            base = {"action_taken": "blocked"}
            evaluation = evaluate_intent_alignment_v2(
                context=normalized, base_result=base, policy=self._policy
            )
            timestamp, clamped = self._allocate_timestamp()
            if timestamp < suspension["expires_at_ms"]:
                raise StrictReceiptCoordinatorV2Error("suspension has not expired")
            index = next(
                (
                    index
                    for index, item in enumerate(self._prior_actions)
                    if item["receipt_hash"] == receipt_hash
                ),
                None,
            )
            if index is None:
                raise StrictReceiptCoordinatorV2Error(
                    "suspended action summary is missing"
                )
            source_hash = v2_canonical_hash(
                {
                    "schema": "obsvr-strict-timeout-evidence-v2",
                    "tenant_id": self._tenant_id,
                    "session_id": self._session_id,
                    "suspended_receipt_hash": receipt_hash,
                    "expires_at_ms": suspension["expires_at_ms"],
                }
            )
            receipt = sign_resolution_v2(
                prior=pending["receipt"],
                method="expired",
                principal_id="obsvr:strict-receipt-coordinator",
                source_hash=source_hash,
                tenant_id=self._tenant_id,
                session_id=self._session_id,
                sequence=self._sequence + 1,
                timestamp=timestamp,
                clamped=clamped,
                previous_hash=self._last_receipt_hash,
                sdk_version=self._sdk_version,
                signer=self._signer,
                include_public_key=self._include_public_key,
                context=normalized,
                base_result=base,
                evaluation=evaluation,
                policy_version=policy_version,
                rule_ids=rule_ids,
            )
            prepared = self._prepared_state.prepare(
                fingerprint=fingerprint,
                receipt_hash=receipt["receipt_hash"],
                kind="timeout",
                value=receipt,
                commit=lambda: self._commit_resolution(receipt, receipt_hash, index),
            )
            return copy.deepcopy(prepared)

    def commit_prepared(self, token: str, receipt_hash: str) -> Any:
        with self._lock:
            self._ensure_process()
            return copy.deepcopy(self._prepared_state.commit(token, receipt_hash))

    def abort_prepared(self, token: str, receipt_hash: str, capability: Any) -> None:
        with self._lock:
            self._ensure_process()
            self._prepared_state.abort(token, receipt_hash, capability)

    def freeze_prepared(
        self,
        token: str,
        receipt_hash: str,
        reason: str = "transport_ambiguous",
    ) -> None:
        with self._lock:
            self._ensure_process()
            self._prepared_state.freeze(token, receipt_hash, reason)

    def inspect_state(self) -> Dict[str, Any]:
        with self._lock:
            self._ensure_process()
            return {
                "tenant_id": self._tenant_id,
                "session_id": self._session_id,
                "sequence": self._sequence,
                "head_receipt_hash": self._last_receipt_hash,
                **self._prepared_state.inspect(),
            }

    def _pending(self, receipt_hash: str) -> Dict[str, Any]:
        pending = self._suspended.get(receipt_hash)
        if pending is None:
            raise StrictReceiptCoordinatorV2Error("suspended receipt is not known")
        if receipt_hash in self._resolved:
            raise StrictReceiptCoordinatorV2Error("suspension is already resolved")
        return pending

    def _allocate_timestamp(self) -> tuple[int, bool]:
        observed = v2_integer(self._clock(), "clock")
        timestamp = max(observed, self._last_timestamp or 0)
        return timestamp, observed < timestamp

    def _commit_decision(
        self,
        result: Dict[str, Any],
        context: Dict[str, Any],
        base: Dict[str, Any],
    ) -> None:
        receipt = result["receipt"]
        self._advance(receipt)
        self._action_ids.add(receipt["body"]["action"]["action_id"])
        self._prior_actions.append(
            {
                "sequence": receipt["body"]["sequence"],
                "kind": receipt["body"]["action"]["kind"],
                "name": receipt["body"]["action"]["name"],
                "outcome": receipt["body"]["evaluation"]["outcome"],
                "receipt_hash": receipt["receipt_hash"],
                "data_classifications": list(context["action"]["data_classifications"]),
            }
        )
        if "suspension" in receipt["body"]:
            self._suspended[receipt["receipt_hash"]] = {
                "receipt": copy.deepcopy(receipt),
                "context": copy.deepcopy(context),
                "base_result": copy.deepcopy(base),
            }
        request_id = receipt["body"].get("suspension", {}).get("approval_request_id")
        if request_id is not None:
            self._approval_requests[request_id] = receipt["receipt_hash"]

    def _commit_resolution(
        self,
        receipt: Dict[str, Any],
        suspended_hash: str,
        index: int,
    ) -> None:
        classifications = self._prior_actions[index]["data_classifications"]
        self._advance(receipt)
        self._prior_actions[index] = {
            "sequence": receipt["body"]["sequence"],
            "kind": receipt["body"]["action"]["kind"],
            "name": receipt["body"]["action"]["name"],
            "outcome": receipt["body"]["evaluation"]["outcome"],
            "receipt_hash": receipt["receipt_hash"],
            "data_classifications": list(classifications),
        }
        self._prior_actions.sort(key=lambda item: item["sequence"])
        self._resolved.add(suspended_hash)

    def _advance(self, receipt: Dict[str, Any]) -> None:
        self._sequence = receipt["body"]["sequence"]
        self._last_receipt_hash = receipt["receipt_hash"]
        self._last_timestamp = receipt["body"]["timestamp_ms"]

    def _ensure_process(self) -> None:
        if v2_integer(self._pid_source(), "pid") != self._owner_pid:
            raise StrictReceiptCoordinatorV2Error(
                "strict v2 coordinator cannot cross a process boundary"
            )


__all__ = ["StrictReceiptCoordinatorV2", "StrictReceiptCoordinatorV2Error"]

File: sdk-typescript/src/governance/strict-receipt-coordinator-v2.ts
import { randomUUID } from 'node:crypto';
import type { DeviceSigner } from '../proxy/device-identity.js';
import {
  type IntentAlignmentV2Result, type IntentPolicyV2Document, type IntentV2BaseResult,
} from '../policy/intent-alignment-v2.js';
import type { ActionContextV2Document, PriorActionV2Input } from './action-context-v2.js';
import type { AarmOutcome } from './aarm-outcome.js';
import {
  buildCoordinatorContextV2, buildIntentPolicyV2, contextInputFromV2Document,
  decisionV2Fingerprint, evaluateV2, resolutionV2Fingerprint, signDecisionV2,
  signResolutionV2, timeoutV2Fingerprint, v2CanonicalHash, v2Clone, v2Hash,
  v2Integer, v2Text, validateDeferredChangesV2,
} from './strict-receipt-coordinator-v2-support.js';
import type {
  PreparedDecisionV2, PreparedResolutionV2, StrictCoordinatorV2StateInspection,
  StrictDecisionV2Input, StrictDecisionV2Result, StrictReceiptCoordinatorV2Options,
  StrictResolutionV2Input, StrictTimeoutV2Input,
} from './strict-receipt-coordinator-v2-types.js';
import { trustedApprovalResult, type StrictApprovalVerifier } from './strict-receipt-coordinator-support.js';
import { type StrictReceiptV2Envelope, strictReceiptV2KeyId } from './strict-receipt-v2.js';
import {
  PreparedReceiptState, type DefinitiveNoStore,
} from './strict-receipt-prepared-state.js';

export type {
  PreparedDecisionV2, PreparedResolutionV2, StrictCoordinatorV2ContextInput,
  StrictCoordinatorV2StateInspection, StrictDecisionV2Input, StrictDecisionV2Result,
  StrictReceiptCoordinatorV2Options, StrictResolutionV2Input, StrictTimeoutV2Input,
} from './strict-receipt-coordinator-v2-types.js';

const FINAL = new Set<AarmOutcome>(['ALLOW', 'DENY', 'MODIFY']);
export interface PendingV2 {
  receipt: StrictReceiptV2Envelope;
  context: ActionContextV2Document;
  baseResult: IntentV2BaseResult;
}

export class StrictReceiptCoordinatorV2Error extends Error {
  constructor(message: string) { super(message); this.name = 'StrictReceiptCoordinatorV2Error'; }
}

export class StrictReceiptCoordinatorV2 {
  protected readonly signer: DeviceSigner;
  private readonly policy: IntentPolicyV2Document;
  protected readonly sdkVersion: string;
  protected readonly tenantId: string;
  protected readonly sessionId: string;
  private readonly clock: () => number;
  private readonly deferTtlMs: number;
  private readonly approvalVerifier: StrictApprovalVerifier;
  private readonly includePublicKey: boolean;
  private readonly pidSource: () => number;
  protected readonly ownerPid: number;
  protected sequence = 0;
  protected lastReceiptHash: string | null = null;
  protected lastTimestamp: number | null = null;
  protected priorActions: PriorActionV2Input[] = [];
  protected readonly suspended = new Map<string, PendingV2>();
  protected readonly resolved = new Set<string>();
  protected readonly actionIds = new Set<string>();
  protected readonly approvalRequests = new Map<string, string>();
  private readonly preparedState: PreparedReceiptState;

  constructor(options: StrictReceiptCoordinatorV2Options) {
    if (options.sdk_language !== 'typescript') throw new StrictReceiptCoordinatorV2Error('sdk_language must be typescript');
    this.signer = options.signer;
    this.policy = buildIntentPolicyV2(options.policy);
    this.sdkVersion = v2Text(options.sdk_version, 'sdk_version');
    this.tenantId = v2Text(options.tenant_id, 'tenant_id');
    this.sessionId = v2Text(options.session_id, 'session_id');
    this.clock = options.clock;
    this.deferTtlMs = v2Integer(options.defer_ttl_ms, 'defer_ttl_ms');
    if (this.deferTtlMs === 0) throw new StrictReceiptCoordinatorV2Error('defer_ttl_ms must be positive');
    if (typeof options.approval_verifier !== 'function') throw new StrictReceiptCoordinatorV2Error('approval_verifier must be a function');
    this.approvalVerifier = options.approval_verifier;
    this.includePublicKey = options.include_public_key ?? true;
    this.pidSource = options.pid ?? (() => process.pid);
    this.ownerPid = v2Integer(this.pidSource(), 'pid');
    this.preparedState = new PreparedReceiptState(options.prepared_token_factory ?? (() => randomUUID()));
    strictReceiptV2KeyId(this.signer.rawPublicKey);
  }

  prepareDecision(input: StrictDecisionV2Input): PreparedDecisionV2 {
    this.ensureProcess();
    const actionId = v2Text(input.action_id, 'action_id');
    const fingerprint = decisionV2Fingerprint(input, this.tenantId, this.sessionId);
    if (this.actionIds.has(actionId)) throw new StrictReceiptCoordinatorV2Error('action_id is already committed');
    const retry = this.preparedState.retry<StrictDecisionV2Result>(fingerprint, 'decision');
    if (retry) return v2Clone(retry);
    const baseResult = { ...input.base_result };
    const context = buildCoordinatorContextV2(input.context, this.sessionId, this.priorActions);
    const evaluation = evaluateV2(context, baseResult, this.policy);
    const { timestamp, clamped } = this.allocateTimestamp();
    const receipt = signDecisionV2({
      action_id: actionId, tenant_id: this.tenantId, session_id: this.sessionId,
      sequence: this.sequence + 1, timestamp, clamped, previous_hash: this.lastReceiptHash,
      sdk_version: this.sdkVersion, signer: this.signer, include_public_key: this.includePublicKey,
      context, base_result: baseResult, evaluation, policy_version: input.policy_version,
      rule_ids: input.rule_ids, defer_ttl_ms: this.deferTtlMs,
    });
    const requestId = receipt.body.suspension?.approval_request_id;
    if (requestId !== undefined && this.approvalRequests.has(requestId)) {
      throw new StrictReceiptCoordinatorV2Error('approval_request_id is already pending');
    }
    const result = { evaluation, receipt };
    return v2Clone(this.preparedState.prepare({
      fingerprint, receipt_hash: receipt.receipt_hash, kind: 'decision', value: result,
      commit: () => this.commitDecision(result, context, baseResult),
    }));
  }

  prepareResolution(input: StrictResolutionV2Input): PreparedResolutionV2 {
    this.ensureProcess();
    if ((input.method as string) === 'expired') {
      throw new StrictReceiptCoordinatorV2Error('expired suspensions must use prepareTimeout');
    }
    const fingerprint = resolutionV2Fingerprint(input, this.tenantId, this.sessionId);
    const retry = this.preparedState.retry<StrictReceiptV2Envelope>(fingerprint, 'resolution');
    if (retry) return v2Clone(retry);
    const receiptHash = v2Hash(input.suspended_receipt_hash, 'suspended_receipt_hash');
    const pending = this.pending(receiptHash);
    const suspension = pending.receipt.body.suspension!;
    const context = buildCoordinatorContextV2(input.context, this.sessionId, this.priorActions);
    const baseResult = { ...input.base_result };
    const evaluation = evaluateV2(context, baseResult, this.policy);
    const { timestamp, clamped } = this.allocateTimestamp();
    const evidence = this.validateResolution(input, pending, context, baseResult, evaluation, timestamp);
    const index = this.priorActions.findIndex((item) => item.receipt_hash === receiptHash);
    if (index < 0) throw new StrictReceiptCoordinatorV2Error('suspended action summary is missing');
    const receipt = signResolutionV2({
      prior: pending.receipt, method: input.method, principal_id: evidence.principalId,
      source_hash: evidence.sourceHash, tenant_id: this.tenantId, session_id: this.sessionId,
      sequence: this.sequence + 1, timestamp, clamped, previous_hash: this.lastReceiptHash,
      sdk_version: this.sdkVersion, signer: this.signer, include_public_key: this.includePublicKey,
      context, base_result: baseResult, evaluation, policy_version: input.policy_version,
      rule_ids: input.rule_ids,
    });
    return v2Clone(this.preparedState.prepare({ fingerprint, receipt_hash: receipt.receipt_hash,
      kind: 'resolution', value: receipt, commit: () => this.commitResolution(receipt, receiptHash, index) }));
  }

  prepareTimeout(input: StrictTimeoutV2Input): PreparedResolutionV2 {
    this.ensureProcess();
    const fingerprint = timeoutV2Fingerprint(input, this.tenantId, this.sessionId);
    const retry = this.preparedState.retry<StrictReceiptV2Envelope>(fingerprint, 'timeout');
    if (retry) return v2Clone(retry);
    const receiptHash = v2Hash(input.suspended_receipt_hash, 'suspended_receipt_hash');
    const pending = this.pending(receiptHash);
    const suspension = pending.receipt.body.suspension!;
    const context = buildCoordinatorContextV2(contextInputFromV2Document(pending.context), this.sessionId, this.priorActions);
    const baseResult: IntentV2BaseResult = { action_taken: 'blocked' };
    const evaluation = evaluateV2(context, baseResult, this.policy);
    const { timestamp, clamped } = this.allocateTimestamp();
    if (timestamp < suspension.expires_at_ms) throw new StrictReceiptCoordinatorV2Error('suspension has not expired');
    const index = this.priorActions.findIndex((item) => item.receipt_hash === receiptHash);
    if (index < 0) throw new StrictReceiptCoordinatorV2Error('suspended action summary is missing');
    const sourceHash = v2CanonicalHash({ schema: 'obsvr-strict-timeout-evidence-v2',
      tenant_id: this.tenantId, session_id: this.sessionId, suspended_receipt_hash: receiptHash,
      expires_at_ms: suspension.expires_at_ms });
    const receipt = signResolutionV2({ prior: pending.receipt, method: 'expired',
      principal_id: 'obsvr:strict-receipt-coordinator', source_hash: sourceHash,
      tenant_id: this.tenantId, session_id: this.sessionId, sequence: this.sequence + 1,
      timestamp, clamped, previous_hash: this.lastReceiptHash, sdk_version: this.sdkVersion,
      signer: this.signer, include_public_key: this.includePublicKey, context, base_result: baseResult,
      evaluation, policy_version: input.policy_version, rule_ids: input.rule_ids });
    return v2Clone(this.preparedState.prepare({ fingerprint, receipt_hash: receipt.receipt_hash,
      kind: 'timeout', value: receipt, commit: () => this.commitResolution(receipt, receiptHash, index) }));
  }

  commitPrepared(token: string, receiptHash: string): StrictDecisionV2Result | StrictReceiptV2Envelope {
    this.ensureProcess();
    return v2Clone(this.preparedState.commit(token, receiptHash));
  }
  abortPrepared(token: string, receiptHash: string, capability: DefinitiveNoStore): void {
    this.ensureProcess(); this.preparedState.abort(token, receiptHash, capability);
  }
  freezePrepared(token: string, receiptHash: string, reason = 'transport_ambiguous'): void {
    this.ensureProcess(); this.preparedState.freeze(token, receiptHash, reason);
  }
  inspectState(): StrictCoordinatorV2StateInspection {
    this.ensureProcess();
    return { tenant_id: this.tenantId, session_id: this.sessionId, sequence: this.sequence,
      head_receipt_hash: this.lastReceiptHash, ...this.preparedState.inspect() };
  }

  private validateResolution(
    input: StrictResolutionV2Input, pending: PendingV2, context: ActionContextV2Document,
    base: IntentV2BaseResult, evaluation: IntentAlignmentV2Result, timestamp: number,
  ): { principalId: string; sourceHash: string } {
    if (!FINAL.has(evaluation.outcome)) throw new StrictReceiptCoordinatorV2Error('resolution evaluation must be final');
    const prior = pending.receipt; const suspension = prior.body.suspension!;
    if (context.agent.agent_id !== prior.body.initiator.agent_id || context.action.kind !== prior.body.action.kind
      || context.action.name !== prior.body.action.name || context.action.arguments_hash !== prior.body.action.arguments_hash
      || context.action.target_hash !== prior.body.action.target_hash) {
      throw new StrictReceiptCoordinatorV2Error('resolution action or initiator does not match suspension');
    }
    const approvalMethod = input.method === 'approval_granted' || input.method === 'approval_denied';
    if ((suspension.type === 'approval') !== approvalMethod && input.method !== 'cancelled') {
      throw new StrictReceiptCoordinatorV2Error('resolution method does not match suspension');
    }
    if (approvalMethod) {
      if (input.resolver_principal_id !== undefined || input.resolution_source_hash !== undefined) {
        throw new StrictReceiptCoordinatorV2Error('approval source must come from approval_evidence');
      }
      if (this.approvalRequests.get(suspension.approval_request_id!) !== prior.receipt_hash) {
        throw new StrictReceiptCoordinatorV2Error('approval request belongs to another suspension');
      }
      const expected = { request_id: suspension.approval_request_id!, action_hash: suspension.approval_action_hash!,
        decision: input.method === 'approval_granted' ? 'granted' as const : 'denied' as const, current_time_ms: timestamp };
      const trusted = trustedApprovalResult(this.approvalVerifier(input.approval_evidence, expected), expected, suspension.expires_at_ms);
      this.validateFinal(input.method, prior, base, evaluation, timestamp);
      return { principalId: trusted.principalId, sourceHash: trusted.sourceHash };
    }
    if (input.approval_evidence !== undefined) throw new StrictReceiptCoordinatorV2Error('approval_evidence requires an approval method');
    if (suspension.type === 'context') validateDeferredChangesV2({ original_context: pending.context,
      current_context: context, original_base: pending.baseResult, current_base: base,
      required_fields: suspension.required_fields });
    const principalId = v2Text(input.resolver_principal_id, 'resolver_principal_id');
    const sourceHash = v2Hash(input.resolution_source_hash, 'resolution_source_hash');
    this.validateFinal(input.method, prior, base, evaluation, timestamp);
    return { principalId, sourceHash };
  }

  private validateFinal(method: StrictResolutionV2Input['method'], prior: StrictReceiptV2Envelope,
    base: IntentV2BaseResult, evaluation: IntentAlignmentV2Result, timestamp: number): void {
    const suspension = prior.body.suspension!;
    if ((method === 'approval_granted' || method === 'context_supplied') && timestamp >= suspension.expires_at_ms) {
      throw new StrictReceiptCoordinatorV2Error('authorization cannot occur after suspension expiry');
    }
    if (['approval_denied', 'cancelled'].includes(method) && evaluation.outcome !== 'DENY') {
      throw new StrictReceiptCoordinatorV2Error('resolution method requires DENY');
    }
    if (method === 'approval_granted' && !['ALLOW', 'MODIFY'].includes(evaluation.outcome)) {
      throw new StrictReceiptCoordinatorV2Error('approval_granted requires ALLOW or MODIFY');
    }
    if (evaluation.outcome === 'MODIFY') {
      const effective = v2Hash(base.modified_arguments_hash, 'modified_arguments_hash');
      if (effective === prior.body.action.arguments_hash) throw new StrictReceiptCoordinatorV2Error('effective_arguments_hash must change arguments');
    }
  }

  private pending(hash: string): PendingV2 {
    const value = this.suspended.get(hash);
    if (!value) throw new StrictReceiptCoordinatorV2Error('suspended receipt is not known');
    if (this.resolved.has(hash)) throw new StrictReceiptCoordinatorV2Error('suspension is already resolved');
    return value;
  }
  private allocateTimestamp(): { timestamp: number; clamped: boolean } {
    const observed = v2Integer(this.clock(), 'clock'); const timestamp = Math.max(observed, this.lastTimestamp ?? 0);
    return { timestamp, clamped: observed < timestamp };
  }
  protected commitDecision(result: StrictDecisionV2Result, context: ActionContextV2Document, base: IntentV2BaseResult): void {
    const receipt = result.receipt; this.advance(receipt); this.actionIds.add(receipt.body.action.action_id);
    this.priorActions.push({ sequence: receipt.body.sequence, kind: receipt.body.action.kind,
      name: receipt.body.action.name, outcome: receipt.body.evaluation.outcome,
      receipt_hash: receipt.receipt_hash, data_classifications: [...context.action.data_classifications] });
    if (receipt.body.suspension) this.suspended.set(receipt.receipt_hash,
      { receipt: v2Clone(receipt), context: v2Clone(context), baseResult: { ...base } });
    const requestId = receipt.body.suspension?.approval_request_id;
    if (requestId !== undefined) this.approvalRequests.set(requestId, receipt.receipt_hash);
  }
  protected commitResolution(receipt: StrictReceiptV2Envelope, hash: string, index: number): void {
    const classifications = this.priorActions[index]!.data_classifications; this.advance(receipt);
    this.priorActions[index] = { sequence: receipt.body.sequence, kind: receipt.body.action.kind,
      name: receipt.body.action.name, outcome: receipt.body.evaluation.outcome,
      receipt_hash: receipt.receipt_hash, data_classifications: [...classifications] };
    this.priorActions.sort((left, right) => left.sequence - right.sequence); this.resolved.add(hash);
  }
  private advance(receipt: StrictReceiptV2Envelope): void {
    this.sequence = receipt.body.sequence; this.lastReceiptHash = receipt.receipt_hash;
    this.lastTimestamp = receipt.body.timestamp_ms;
  }
  protected ensureProcess(): void {
    if (v2Integer(this.pidSource(), 'pid') !== this.ownerPid) {
      throw new StrictReceiptCoordinatorV2Error('strict v2 coordinator cannot cross a process boundary');
    }
  }
}
